lib/postcode_validation/use_case/validate_address.rb

```diff
@@ -6,42 +6,59 @@ def initialize(address_match_gateway:, logger: nil)
         @format_validator = FormatValidator.new
         @address_match_gateway = address_match_gateway
         @logger = logger
+        @errors = []
       end
 
       def execute(postcode:, country:)
-        @postcode = postcode
-        @country = country
-        { valid?: valid_postcode? }
+        ensure_country(country)
+        ensure_postcode_format(postcode, country)
+        result = matched_addresses(postcode, country)
+
+        ensure_results(result)
+
+        result_payload(result, postcode)
       rescue PostcodeValidation::Error::RequestError => e
-        on_error(e)
-        { valid?: true }
+        gracefully_handle_error(e)
       end
 
       private
 
       attr_reader :address_match_gateway, :logger, :postcode, :country
 
+      def result_payload(result, postcode)
+        return { valid?: false, reason: @errors } unless @errors.empty?
+
+        result.each do |address|
+          return { valid?: true, reason: 'valid_postcode' } if address.postcode_matches? postcode
```

Comment:
reason: should always be an array. string is liskov incompatible with array which makes this all hell harder to consume in static-typed languages